### PR TITLE
Bind Kestrel options to config by default (#30)

### PR DESCRIPTION
**Pull request by cesarbs**
#30

### CODE REVIEW

File: src/Microsoft.AspNetCore/WebHost.cs [hunk 1]
```diff
@@ -148,6 +150,7 @@ public static IWebHostBuilder CreateDefaultBuilder(string[] args)
 
                     if (env.IsDevelopment())
                     {
+                        Console.WriteLine(env.ApplicationName);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/WebHost.cs]: I assume this is just left over.

File: src/Microsoft.AspNetCore/Microsoft.AspNetCore.csproj [hunk 2]
```diff
@@ -11,7 +11,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-   <PackageReference Include="@(MetaPackagePackageReference)" />
+    <PackageReference Include="@(MetaPackagePackageReference)" />
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/Microsoft.AspNetCore.csproj]: Undo this change?

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore
+{
+    /// <summary>
+    /// Binds Kestrel configuration.
+    /// </summary>
+    public class KestrelServerOptionsSetup : IConfigureOptions<KestrelServerOptions>
+    {
+        private IServiceProvider _services;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: This should be `IConfiguration`, not `IServiceProvider`

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+        /// <summary>
+        /// Creates a new instance of <see cref="KestrelServerOptionsSetup"/>.
+        /// </summary>
+        /// <param name="services">An <seealso cref="IServiceProvider"/> instance.</param>
+        public KestrelServerOptionsSetup(IServiceProvider services)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: See comment above. Same thing.

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        /// </summary>
+        /// <param name="certificateConfiguration">An <seealso cref="IConfiguration"/> with information about a certificate source.</param>
+        /// <param name="password">The certificate password, in case it's being loaded from a file.</param>
+        /// <returns>The loaded certificates.</returns>
+        public static X509Certificate2 Load(IConfiguration certificateConfiguration, string password = null)
+        {
+            var sourceKind = certificateConfiguration.GetValue<string>("Source");
+
+            CertificateSource certificateSource;
+            switch (sourceKind)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Is this supposed to be case sensitive?

Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: cc @javiercn

File: src/Microsoft.AspNetCore/Microsoft.AspNetCore.csproj (hunk 2 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore/Microsoft.AspNetCore.csproj]: Spacing was wrong.

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        /// Loads all certificates specified in an <seealso cref="IConfiguration"/>.
+        /// </summary>
+        /// <param name="configurationRoot">The root <seealso cref="IConfiguration"/>.</param>
+        /// <returns>
+        /// A dictionary mapping certificate names to loaded certificates.
+        /// </returns>
+        public static Dictionary<string, X509Certificate2> LoadAll(IConfiguration configurationRoot)
+        {
+            var certificates = configurationRoot.GetSection("Certificates");
+            var loadedCertificates = new Dictionary<string, X509Certificate2>();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: This looks like it could be a single LINQ statement.

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                BindEndPoint(options, endPoint, certificates);
+            }
+        }
+
+        private void BindEndPoint(
+            KestrelServerOptions options,
+            IConfigurationSection endPoint,
+            Dictionary<string, X509Certificate2> certificates)
+        {
+            options.Listen(IPAddress.Parse(endPoint.GetValue<string>("Address")), endPoint.GetValue<int>("Port"), listenOptions =>
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: This needs to be more failure tolerant. TryParse with a decent exception message

File: samples/AppSettings/Program.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+using System;
+using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Http;
+
+namespace AppSettings
+{
+    public class Program
+    {
+        public static void Main(string[] args)
```
Comment (Tratcher) [inline samples/AppSettings/Program.cs]: Why not use the existing sample?

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 5 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: No, avoid treating config as case sensitive.

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+        private class CertificateStoreSource : CertificateSource
+        {
+            public string Subject { get; set; }
+            public string StoreName { get; set; }
+            public string StoreLocation { get; set; }
+
+            public override X509Certificate2 Load()
+            {
+                if (!Enum.TryParse(StoreLocation, true, out StoreLocation storeLocation))
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: named params

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+                {
+                    throw new InvalidOperationException($"Invalid store location: {StoreLocation}");
+                }
+
+                using (var store = new X509Store(StoreName, storeLocation))
+                {
+                    store.Open(OpenFlags.ReadOnly);
+                    var foundCertificate = store.Certificates.Find(X509FindType.FindBySubjectName, Subject, validOnly: false)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Is there an EKU check anywhere?

File: samples/AppSettings/appsettings.json [hunk 11]
```diff
@@ -0,0 +1,61 @@
+{
+  "Kestrel": {
+    "EndPoints": {
+      "Http": {
+        "Address": "127.0.0.1",
+        "Port": 8081
+      },
+      "HttpV6": {
+        "Address": "[::1]",
```
Comment (Tratcher) [inline samples/AppSettings/appsettings.json]: Brackets not required

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            }
+
+            options.Listen(address, port, listenOptions =>
+            {
+                var certificateName = endPoint.GetValue<string>("Certificate");
+
+                X509Certificate2 endPointCertificate = null;
+                if (certificateName != null)
+                {
+                    endPointCertificate = certificates[certificateName];
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: KeyNotFoundException? You can give a better error.

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+                }
+                else
+                {
+                    var certificate = endPoint.GetSection("Certificate");
+
+                    if (certificate.GetChildren().Any())
+                    {
+                        var password = _configurationRoot[$"Kestrel:EndPoints:{endPoint.Key}:Certificate:Password"];
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: is this not already part of the certificate config section? even if it's coming from a different config source it should get merged.

File: samples/AppSettings/Program.cs (hunk 8 above)
Comment (cesarbs) [inline samples/AppSettings/Program.cs]: Settings would apply to all the scenarios there. I want to keep things separate for clarity.

File: samples/SampleApp/SampleApp.csproj [hunk 14]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk.Web">
 
-  <Import Project="..\..\build\common.props" />
+  <Import Project="..\..\build\dependencies.props" />
```
Comment (davidfowl) [inline samples/SampleApp/SampleApp.csproj]: ?

Comment (cesarbs) [inline samples/SampleApp/SampleApp.csproj]: @natemcmaster pointed out we don't normally import `common.props` here, only `dependencies.props`. This was causing issues with user secrets because `common.props` has `GenerateUserSecretsAttribute` set to false.

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+                }
+
+                return certificate;
+            }
+
+            private X509Certificate2 TryLoad(X509KeyStorageFlags flags, out Exception exception)
+            {
+                try
+                {
+                    var loadedCertificate = new X509Certificate2(Path, _password);
```
Comment (javiercn) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: var loadedCertificate = new X509Certificate2(Path, _password, flags);

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+            }
+
+            public string Path { get; set; }
+
+            public override X509Certificate2 Load()
+            {
+                var certificate = TryLoad(X509KeyStorageFlags.DefaultKeySet, out var error)
+                    ?? TryLoad(X509KeyStorageFlags.UserKeySet, out error)
+    #if NETCOREAPP2_0
+                    ?? TryLoad(X509KeyStorageFlags.EphemeralKeySet, out error)
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: @blowdart Is this the proper order to attempt to load the cert? cc @javiercn

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+            private X509Certificate2 TryLoad(X509KeyStorageFlags flags, out Exception exception)
+            {
+                try
+                {
+                    var certificate = new X509Certificate2(Path, _password, flags);
+
+                    if (!HasServerAuthEKU(certificate))
+                    {
+                        throw new InvalidOperationException(
+                            $"The certificate file at '{Path}' does not contain an Enhanced Key Usage field with the OID 1.3.6.1.5.5.7.3.1 (Server Authentication)");
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: @blowdart Is the wording correct?

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+                        .FirstOrDefault();
+
+    #if NET46
+                    store.Close();
+    #endif
+
+                    if (foundCertificate == null)
+                    {
+                        throw new InvalidOperationException(
+                            $"No valid certificate found for {Subject} in store {StoreName} under {StoreLocation}. Check that the certificate contains the Enhanced Key Usage field with the OID 1.3.6.1.5.5.7.3.1 (Server Authentication)");
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: @blowdart Is the wording correct?

**Comment by cesarbs** [comment]
@danroth27 Should the EKU check be mandatory as I've done it here? Or should we add a flag to optionally disable it?

Also: should we actually move the check to Kestrel?

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 10 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Added.

**Comment by danroth27** [comment]
Yeah, if we're going to check the EKUs at all it should be done lower in the stack. Let's remove this check from here.

**Comment by cesarbs** [comment]
Removed EKU check. Needs further discussion.

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+
+                using (var store = new X509Store(StoreName, storeLocation))
+                {
+                    store.Open(OpenFlags.ReadOnly);
+                    var foundCertificate = store.Certificates.Find(X509FindType.FindBySubjectName, Subject, validOnly: false)
+                        .OfType<X509Certificate2>()
+                        .OrderByDescending(certificate => certificate.NotAfter)
+                        .FirstOrDefault();
+
+    #if NET46
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: No needed.

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+            var portValue = endPoint.GetValue<string>("Port");
+
+            IPAddress address;
+            if (!IPAddress.TryParse(addressValue, out address))
+            {
+                throw new InvalidOperationException($"Invalid IP address: {addressValue}");
+            }
+
+            int port;
+            if (!int.TryParse(portValue, out port))
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: culture, style

**Comment by cesarbs** [comment]
@halter73 Added you as a reviewer.

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 10 above)
Comment (blowdart) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Why validOnly: false? Do we want people to be able to bind to expired certs?

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 16 above)
Comment (blowdart) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: What are you trying to do here? @bartonjs because he understands Ephemeral :)

**Comment by blowdart** [comment]
@danroth27 It's not a question of IF we perform an EKU check. We should/must. So if it's removed from here, then where is the corresponding issue in kestrel to perform the check there?

**Comment by danroth27** [comment]
@blowdart Filed https://github.com/aspnet/KestrelHttpServer/issues/1764 to track validating the cert EKU

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 16 above)
Comment (bartonjs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: I read this as:

* Load perphemeral, as MachineKeySet as the PFX says so, or as UserKeySet if it has no opinion on the matter.
* Load perphemeral as UserKeySet (in case the key was MachineKeySet and the current process does not have write access to the machine key store).
* Load ephemeral (in case the current process has no persisted key write access at all, such as low rights with no user profile)

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 10 above)
Comment (bartonjs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: * FindBySubjectName does strstr. If you want exact match you'll need to post-process it.
* store.Certificates returns unique X509Certificate2 objects, and Find returns unique X509Certificate2 objects.  If you would like to reduce Finalizations you should dispose all of the members of store.Certificates (which you need to save as a local, since reading it twice just makes the problem worse).

Similary, if you cascade call Find you should dispose all the previous values.

**Comment by cesarbs** [comment]
📢

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 10 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: @javiercn Since this was originally your code, what are your thoughts here?

Comment (javiercn) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: @blowdart self-signed certificates. I would change that to HostingEnvironment.IsDevelopment() if its available

Comment (javiercn) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Another option is to just load the certificates and filter using Linq, or have an extra configuration value per certificate that indicates if its ok to load invalid (so that you can again load self-signed certificates for development)

Yet another option is to make the check true and force the user to put the certificate (in case its self-signed) in the trusted certificate store for .NET. (Which varies across platforms)

Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Talked to @danroth27 and we decided to go with a config key (`AllowInvalid`).

**Comment by cesarbs** [comment]
@bartonjs Can you review the changes I made to loading the certs from the store? Disposing & validating subject further.

**Comment by bartonjs** [comment]
Cert/store loading looks good to me.  You might consider plumbing your EKU check to a `.Where` so that you don't pick a new-improper cert over a slightly-older-but-valid cert.

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+                    X509Certificate2Collection foundCertificates = null;
+
+                    try
+                    {
+                        store.Open(OpenFlags.ReadOnly);
+                        storeCertificates = store.Certificates;
+                        foundCertificates = storeCertificates.Find(X509FindType.FindBySubjectName, Subject, validOnly: !AllowInvalid);
+                        var foundCertificate = foundCertificates
+                            .OfType<X509Certificate2>()
+                            .Where(certificate => certificate.Subject.Equals(Subject, StringComparison.Ordinal))
```
Comment (danroth27) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Why do we need this where clause when we already did a find by subject name?

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+                    if (!certificates.TryGetValue(certificateName, out endPointCertificate))
+                    {
+                        throw new InvalidOperationException($"No certificate named {certificateName} found in configuration");
+                    }
+                }
+                else
+                {
+                    var certificate = endPoint.GetSection("Certificate");
+
+                    if (certificate.GetChildren().Any())
```
Comment (danroth27) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: What happens if the Certificate property isn't there at all on the endpoint? Would this nullref?

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+            if (!int.TryParse(portValue, out port))
+            {
+                throw new InvalidOperationException($"Invalid port: {portValue}");
+            }
+
+            options.Listen(address, port, listenOptions =>
+            {
+                var certificateName = endPoint.GetValue<string>("Certificate");
+
+                X509Certificate2 endPointCertificate = null;
```
Comment (danroth27) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: It would be nice if this logic for dealing with a property that could be a certificate name or an inline certificate was exposed as a helper method on the certificate loader so that we can use it in other places where we want the same behavior (like configuring a signing certificate).

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 21 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: @bartonjs pointed out that `FindSubjectByName` does a substring search, so we need this extra check to get an exact string match.

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs (hunk 22 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: No, you get an empty section.

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 21 above)
Comment (bartonjs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Yeah, then he learned he was both right and wrong.  Silly @bartonjs.  You want FindByDistinguishedSubjectName, that leaves all of the RDN keys in place, and matches on String.Equals.  So it does this exact check for you, and you can then drop the Where.

(FindBySubjectName drops all the tags out, for... reasons?, and also doesn't reverse the fields.  So where you want to match "CN=localhost, OU=fun times, C=ZZ" it matches "ZZ, fun times, localhost")

FindByDistinguishedName matches IgnoreCase for some reason, so if you want to be case-respecting you might want to leave this here after all.

Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: When I use `FindBySubjectDistinguishedName`, I have to change my settings to `CN=localhost` to load my certificate (and I guess that's what you meant by the other option dropping the tags).

@danroth27 Do we want this or should we keep the previous behavior?

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 16 above)
Comment (blowdart) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: "perphemeral"? 🤣 

Is there an advantage in trying "perphemeral" first, if ephermeral is there? There's no real need to persist the private key anyhere.

**Comment by cesarbs** [comment]
@muratg Adding you as a reviewer since this will require preview1 approval.

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 21 above)
Comment (danroth27) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: We discussed this offline and we agreed we want to use the distinguished subject name (i.e. CN=localhost)

**Comment by muratg** [comment]
@cesarbs will you send a PR to the preview1 branch as well?  Also, when is the user-secrets work be done?

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs (hunk 23 above)
Comment (danroth27) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: Some examples of how we would like to reuse the cert loading logic: https://github.com/aspnet/MetaPackages/issues/30#issuecomment-297527522

**Comment by cesarbs** [comment]
@muratg I've rebased this to `rel/2.0.0-preview1`. Once it's merged I'll merge it to `dev`.

**Comment by cesarbs** [comment]
@muratg User secrets working depends on https://github.com/dotnet/cli/pull/6356.

**Comment by cesarbs** [comment]
@JunTaoLuo @Tratcher Can you guys review the functional tests I've just added?

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 16 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: I'll put the call with `EphemeralKeySet` first then, with a fallback to `UserKeySet` if that fails. Does that sound right?

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+                var certificate = TryLoad(X509KeyStorageFlags.DefaultKeySet, out var error)
+                    ?? TryLoad(X509KeyStorageFlags.UserKeySet, out error)
+    #if NETCOREAPP2_0
+                    ?? TryLoad(X509KeyStorageFlags.EphemeralKeySet, out error)
+    #endif
+                    ;
+
+                if (error != null)
+                {
+                    throw error;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Rethrowing this exception looses the original stack trace. You should wrap it or use ExceptionDispatchInfo to capture and rethrow it.

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+        private void BindEndPoint(
+            KestrelServerOptions options,
+            IConfigurationSection endPoint,
+            Dictionary<string, X509Certificate2> certificates)
+        {
+            var addressValue = endPoint.GetValue<string>("Address");
+            var portValue = endPoint.GetValue<string>("Port");
+
+            IPAddress address;
+            if (!IPAddress.TryParse(addressValue, out address))
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: Wait, I can't bind to * or localhost? That's a non-starter.

Comment (Tratcher) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: https://github.com/aspnet/MetaPackages/issues/73
